### PR TITLE
Rework markup link rendering

### LINKED ISSUE #26548: markdown image links are wrong if repo is called 'src'
### Description

Ack, this is an edge case.

We have a repo called src, and amazingly all links to images in markdown are wrong (404).

for example:

document: https://host/org/src/src/commit/...
actual img src: https://host/org/media/src/commit/...
expected img src: https://host/org/src/media/commit/...

It looks like the the first, instead of the second. occurrence of "src" is replaced by "media".

[try.gitea.io](https://try.gitea.io/rmie/src/src/branch/main/README.md) 

[suspected location](https://github.com/go-gitea/gitea/blob/1432d4eab94814db267577ec911a3375a1269fc0/modules/markup/html.go#L412)

### Gitea Version

1.19.3

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

docker

### Database

None

### PR DESCRIPTION
**Pull request by KN4CK3R**
Fixes #26548

This PR refactors the rendering of markup links. The old code uses `strings.Replace` to change some urls while the new code uses more context to decide which link should be generated.

The added tests should ensure the same output for the old and new behaviour (besides the bug).

We may need to refactor the rendering a bit more to make it clear how the different helper methods render the input string. There are lots of options (resolve links / images / mentions / git hashes / emojis / ...) but you don't really know what helper uses which options. For example, we currently support images in the user description which should not be allowed I think:

<details>
  <summary>Profile</summary> 

https://try.gitea.io/KN4CK3R
 grafik

</details>

### CODE REVIEW

File: modules/markup/html.go [hunk 1]
```diff
@@ -390,35 +385,38 @@ func visitNode(ctx *RenderContext, procs, textProcs []processor, node *html.Node
...
 				}
-				if len(attr.Val) > 0 && !isLinkStr(attr.Val) && !strings.HasPrefix(attr.Val, "data:image/") {
-					prefix := ctx.URLPrefix
+				if len(attr.Val) > 0 && !IsLinkStr(attr.Val) && !strings.HasPrefix(attr.Val, "data:image/") {
+					var base string
 					if ctx.IsWiki {
-						prefix = util.URLJoin(prefix, "wiki", "raw")
+						base = ctx.Links.WikiRawLink()
+					} else if ctx.Links.HasBranchInfo() {
+						base = ctx.Links.MediaLink()
```
Comment (KN4CK3R) [inline modules/markup/html.go]: We mix `/raw/` and `/media/`. Only `/media/` should be used I think to support LFS images.

File: modules/markup/html.go [hunk 2]
```diff
@@ -751,16 +745,18 @@ func shortLinkProcessorFull(ctx *RenderContext, node *html.Node, noLink bool) {
 				link = url.PathEscape(link)
 			}
 		}
-		urlPrefix := ctx.URLPrefix
 		if image {
+			var base string
+			if ctx.IsWiki {
+				base = ctx.Links.WikiRawLink()
+			} else if ctx.Links.HasBranchInfo() {
+				base = ctx.Links.RawLink()
```
Comment (KN4CK3R) [inline modules/markup/html.go]: Here we use `/raw/` for both paths. Still should be `/media/`.

File: modules/markup/markdown/markdown.go [hunk 3]
```diff
@@ -66,9 +63,6 @@ func (l *limitWriter) Write(data []byte) (int, error) {
 // newParserContext creates a parser.Context with the render context set
 func newParserContext(ctx *markup.RenderContext) parser.Context {
 	pc := parser.NewContext(parser.WithIDs(newPrefixedIDs()))
-	pc.Set(urlPrefixKey, ctx.URLPrefix)
-	pc.Set(isWikiKey, ctx.IsWiki)
-	pc.Set(renderMetasKey, ctx.Metas)
```
Comment (KN4CK3R) [inline modules/markup/markdown/markdown.go:71]: No need to put them into own keys if we already pass the whole render context. Makes the other code more readable.

File: modules/markup/orgmode/orgmode.go [hunk 4]
```diff
@@ -132,50 +130,41 @@ func (Renderer) Render(ctx *markup.RenderContext, input io.Reader, output io.Wri
...
-			lnk = util.URLJoin("wiki", lnk)
+	if len(link) > 0 && !markup.IsLinkStr(link) &&
+		link[0] != '#' && !strings.HasPrefix(link, mailto) {
+		base := r.Ctx.Links.Base
+		switch l.Kind() {
+		case "image", "video":
+			if r.Ctx.IsWiki {
+				base = r.Ctx.Links.WikiRawLink()
+			} else if r.Ctx.Links.HasBranchInfo() {
+				base = r.Ctx.Links.MediaLink()
```
Comment (KN4CK3R) [inline modules/markup/orgmode/orgmode.go]: raw vs media

File: modules/templates/util_render.go [hunk 5]
```diff
@@ -51,9 +43,9 @@ func RenderCommitMessageLink(ctx context.Context, msg, urlPrefix, urlDefault str
 	return template.HTML(msgLines[0])
 }
 
-// RenderCommitMessageLinkSubject renders commit message as a XXS-safe link to
+// RenderCommitMessageLinkSubject renders commit message as a XSS-safe link to
 // the provided default url, handling for special links without email to links.
-func RenderCommitMessageLinkSubject(ctx context.Context, msg, urlPrefix, urlDefault string, metas map[string]string) template.HTML {
+func RenderCommitMessageLinkSubject(ctx context.Context, msg, urlDefault string, metas map[string]string) template.HTML {
```
Comment (KN4CK3R) [inline modules/templates/util_render.go:48]: Removed `urlPrefix` because `RenderCommitMessageSubject` does not render links at all.

**Comment by CaiCandong** [comment]
Thank you very much for your work. I remember another bug that may be relevant to this PR, maybe it can be fixed together.
Bug https://github.com/go-gitea/gitea/issues/25632
Related https://github.com/go-gitea/gitea/issues/15075

**Comment by CaiCandong** [comment]
By tested locally, overall LGTM, but  I need some time to understand the change

**Review by CaiCandong — APPROVED** [review]

File: modules/markup/renderer.go [hunk 6]
```diff
@@ -80,6 +81,36 @@ type RenderContext struct {
...
+	BranchPath string
+	TreePath   string
+}
+
+func (ctx *Links) HasBranchInfo() bool {
+	return ctx.BranchPath != ""
+}
+
+func (ctx *Links) SrcLink() string {
+	return util.URLJoin(ctx.Base, "src", ctx.BranchPath, ctx.TreePath)
```
Comment (wxiaoguang) [inline modules/markup/renderer.go]: `ctx` doesn't seem a good variable name for `Links`.

And .... could it use a two-word variable like `BaseLink`? There are already `ctx.Base` in our web context.

`BaseLink` makes it easier to grep the whole code base (and matches `SrcLink`, etc)

File: services/mailer/mail_release.go [hunk 7]
```diff
@@ -58,9 +58,11 @@ func mailNewRelease(ctx context.Context, lang string, tos []string, rel *repo_mo
 
 	var err error
 	rel.RenderedNote, err = markdown.RenderString(&markup.RenderContext{
-		Ctx:       ctx,
-		URLPrefix: rel.Repo.Link(),
-		Metas:     rel.Repo.ComposeMetas(),
+		Ctx: ctx,
+		Links: markup.Links{
+			Base: rel.Repo.Link(),
```
Comment (wxiaoguang) [inline services/mailer/mail_release.go]: Maybe it should be HTMLURL to be used in the mail

File: modules/markup/renderer.go (hunk 6 above)
Comment (KN4CK3R) [reply modules/markup/renderer.go]: `ctx` is a leftover when I called `Links` `URLContext`. Will change that.

**Comment by wxiaoguang** [comment]
Resolve a conflict caused by my PR 67bc21d2e87f82df29582e4d073842b7adeb780a

<details>

image

</details>

**Comment by wxiaoguang** [comment]
It has been stable for long time. I will do some work (new PRs) without considering conflict with these changes ......

File: routers/web/shared/user/header.go [hunk 8]
```diff
@@ -43,10 +43,8 @@ func PrepareContextForProfileBigAvatar(ctx *context.Context) {
 
 	if len(ctx.ContextUser.Description) != 0 {
 		content, err := markdown.RenderString(&markup.RenderContext{
-			URLPrefix: ctx.Repo.RepoLink,
```
Comment (delvh) [inline routers/web/shared/user/header.go:50]: What happened to this link?

File: modules/markup/markdown/goldmark.go [hunk 9]
```diff
@@ -87,18 +87,17 @@ func (g *ASTTransformer) Transform(node *ast.Document, reader text.Reader, pc pa
...
-				if pc.Get(isWikiKey).(bool) {
-					prefix = giteautil.URLJoin(prefix, "wiki", "raw")
+				var base string
+				if ctx.IsWiki {
+					base = ctx.Links.WikiRawLink()
+				} else if ctx.Links.HasBranchInfo() {
+					base = ctx.Links.MediaLink()
+				} else {
+					base = ctx.Links.Base
 				}
```
Comment (delvh) [inline modules/markup/markdown/goldmark.go]: I've seen this block now often enough.
Can we perhaps move it into its own function?

File: modules/markup/markdown/goldmark.go [hunk 10]
```diff
@@ -143,11 +142,15 @@ func (g *ASTTransformer) Transform(node *ast.Document, reader text.Reader, pc pa
...
 				// special case: this is not a link, a hash link or a mailto:, so it's a
 				// relative URL
-				lnk := string(link)
-				if pc.Get(isWikiKey).(bool) {
-					lnk = giteautil.URLJoin("wiki", lnk)
+
+				var base string
+				if ctx.IsWiki {
+					base = ctx.Links.WikiLink()
+				} else {
```
Comment (delvh) [inline modules/markup/markdown/goldmark.go:140]: No `else if`?

File: modules/markup/markdown/markdown_test.go [hunk 11]
```diff
@@ -556,3 +570,369 @@ foo: bar
...
 }
+
+func TestRenderLinks(t *testing.T) {
+	input := `  space @mention-user  
+/just/a/path.bin
+https://example.com/file.bin
+[local link](file.bin)
+[remote link](https://example.com)
+[[local link|file.bin]]
+[[remote link|https://example.com]]
```
Comment (delvh) [inline modules/markup/markdown/markdown_test.go:581]: What is this syntax?

File: modules/markup/orgmode/orgmode.go [hunk 12]
```diff
@@ -132,50 +130,41 @@ func (Renderer) Render(ctx *markup.RenderContext, input io.Reader, output io.Wri
...
-		link[0] != '#' && !bytes.HasPrefix(link, byteMailto) {
-		lnk := string(link)
-		if r.IsWiki {
-			lnk = util.URLJoin("wiki", lnk)
+	if len(link) > 0 && !markup.IsLinkStr(link) &&
+		link[0] != '#' && !strings.HasPrefix(link, mailto) {
+		base := r.Ctx.Links.Base
+		switch l.Kind() {
+		case "image", "video":
+			if r.Ctx.IsWiki {
```
Comment (delvh) [inline modules/markup/orgmode/orgmode.go]: See above

File: routers/web/org/home.go [hunk 13]
```diff
@@ -44,10 +44,8 @@ func Home(ctx *context.Context) {
 	ctx.Data["Title"] = org.DisplayName()
 	if len(org.Description) != 0 {
 		desc, err := markdown.RenderString(&markup.RenderContext{
-			Ctx:       ctx,
-			URLPrefix: ctx.Repo.RepoLink,
```
Comment (delvh) [inline routers/web/org/home.go:51]: What happened to this link?

File: modules/templates/util_render.go [hunk 14]
```diff
@@ -211,25 +200,9 @@ func ReactionToEmoji(reaction string) template.HTML {
...
-		log.Error("RenderNote: %v", err)
-		return ""
-	}
-	return template.HTML(fullMessage)
-}
-
 func RenderMarkdownToHtml(ctx context.Context, input string) template.HTML { //nolint:revive
 	output, err := markdown.RenderString(&markup.RenderContext{
-		Ctx:       ctx,
-		URLPrefix: setting.AppSubURL,
```
Comment (delvh) [inline modules/templates/util_render.go:232]: What happened to this link?

File: routers/web/org/home.go (hunk 13 above)
Comment (KN4CK3R) [reply routers/web/org/home.go:51]: The description of an org can't have a repo context and `ctx.Repo.RepoLink` is always empty here.

File: modules/templates/util_render.go (hunk 14 above)
Comment (KN4CK3R) [reply modules/templates/util_render.go:232]: `markdown.RenderString` does not use this field and setting it just creates confusion.

File: routers/web/shared/user/header.go (hunk 8 above)
Comment (KN4CK3R) [reply routers/web/shared/user/header.go:50]: There is no repository in the user/org description context and `RepoLink` is always an empty string.

File: modules/markup/markdown/goldmark.go (hunk 10 above)
Comment (KN4CK3R) [reply modules/markup/markdown/goldmark.go:140]: no

File: modules/markup/markdown/markdown_test.go (hunk 11 above)
Comment (KN4CK3R) [reply modules/markup/markdown/markdown_test.go:581]: Should be the wiki style syntax. Nothing new added in this PR if that's your question.
See here for example:
https://github.com/go-gitea/gitea/blob/d95102d650673165785a05d1cd07a75a3f8763cb/modules/markup/markdown/markdown_test.go#L103-L105

File: routers/web/repo/commit.go [hunk 15]
```diff
@@ -370,9 +374,21 @@ func Diff(ctx *context.Context) {
...
 		ctx.Data["NoteAuthor"] = user_model.ValidateCommitWithEmail(ctx, note.Commit)
+		ctx.Data["NoteRendered"], err = markup.RenderCommitMessage(&markup.RenderContext{
+			Links: markup.Links{
+				Base:       ctx.Repo.RepoLink,
+				BranchPath: path.Join("commit", util.PathEscapeSegments(commitID)),
+			},
+			Metas:   ctx.Repo.Repository.ComposeMetas(ctx),
+			GitRepo: ctx.Repo.GitRepo,
+			Ctx:     ctx,
+		}, template.HTMLEscapeString(string(charset.ToUTF8WithFallback(note.Message))))
```
Comment (wxiaoguang) [inline routers/web/repo/commit.go:387]: ~~`template.HTMLEscapeString(` doesn't seem right here.~~

Hmm ... OK, it seems that  it's intentionally to escape the HTML tags.

**Review by wxiaoguang — APPROVED** [review]
Big enough and good enough

File: templates/repo/branch/list.tmpl [hunk 16]
```diff
@@ -25,7 +25,7 @@
...
+								<p class="info gt-df gt-ac gt-my-2">{{svg "octicon-git-commit" 16 "gt-mr-2"}}<a href="{{.RepoLink}}/commit/{{PathEscape .DefaultBranchBranch.DBBranch.CommitID}}">{{ShortSha .DefaultBranchBranch.DBBranch.CommitID}}</a> · <span class="commit-message">{{RenderCommitMessage $.Context .DefaultBranchBranch.DBBranch.CommitMessage (.Repository.ComposeMetas ctx)}}</span> · {{.locale.Tr "org.repo_updated"}} {{TimeSince .DefaultBranchBranch.DBBranch.CommitTime.AsTime ctx.Locale}}{{if .DefaultBranchBranch.DBBranch.Pusher}} &nbsp;{{template "shared/user/avatarlink" dict "user" .DefaultBranchBranch.DBBranch.Pusher}}{{template "shared/user/namelink" .DefaultBranchBranch.DBBranch.Pusher}}{{end}}</p>
```
Comment (wxiaoguang) [inline templates/repo/branch/list.tmpl]: Oops, merge conflict: `.locale.Tr`

**Pull request merged**